adapters/intertech/intertech.go

```diff
@@ -0,0 +1,340 @@
+			"Content-Type":    "application/json;charset=utf-8",
+			"Accept":          "application/json",
+		})
+	}
+
+	return headers
+}
+
+func addNonEmptyHeaders(headers *http.Header, headerValues map[string]string) {
+	for key, value := range headerValues {
+		if len(value) > 0 {
+			headers.Add(key, value)
+		}
+	}
+}
+
+// Request is in shared memory, so we have to make a shallow copy for further modification (imp is already a shallow copy)
+func splitRequestDataByImp(request *openrtb2.BidRequest, imp openrtb2.Imp) openrtb2.BidRequest {
+	requestCopy := *request
+	requestCopy.Imp = []openrtb2.Imp{imp}
+
+	return requestCopy
+}
+
+func getIntertechPlacementId(imp openrtb2.Imp) (*intertechPlacementID, error) {
+	var ext adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &ext); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext", imp.ID),
+		}
+	}
+
+	var intertechExt openrtb_ext.ExtImpIntertech
+	if err := json.Unmarshal(ext.Bidder, &intertechExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("imp %s: unable to unmarshal ext.bidder: %v", imp.ID, err),
+		}
+	}
+
+	placementID, err := mapExtToPlacementID(intertechExt)
+	if err != nil {
+		return nil, err
+	}
+
+	return placementID, nil
+}
+
+func mapExtToPlacementID(intertechExt openrtb_ext.ExtImpIntertech) (*intertechPlacementID, error) {
+	var placementID intertechPlacementID
+
+	if len(intertechExt.PlacementID) == 0 {
+		placementID.ImpID = strconv.Itoa(int(intertechExt.ImpID))
+		placementID.PageID = strconv.Itoa(int(intertechExt.PageID))
+		return &placementID, nil
+	}
+
+	idParts := strings.Split(intertechExt.PlacementID, "-")
+
+	numericIdParts := []string{}
+
+	for _, idPart := range idParts {
+		if _, err := strconv.Atoi(idPart); err == nil {
+			numericIdParts = append(numericIdParts, idPart)
+		}
+	}
+
+	if len(numericIdParts) < 2 {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("invalid placement id, it must contain two parts: %s", intertechExt.PlacementID),
+		}
+	}
+
+	placementID.ImpID = numericIdParts[len(numericIdParts)-1]
+	placementID.PageID = numericIdParts[len(numericIdParts)-2]
+
+	return &placementID, nil
+}
+
+func modifyImp(imp *openrtb2.Imp) error {
+	if imp.Banner != nil {
+		banner, err := modifyBanner(*imp.Banner)
+		if banner != nil {
+			imp.Banner = banner
+		}
+		return err
+	}
+
+	if imp.Native != nil {
+		return nil
+	}
+
+	return &errortypes.BadInput{
+		Message: fmt.Sprintf("Unsupported format. Intertech only supports banner and native types. Ignoring imp id #%s", imp.ID),
+	}
+}
+
+func modifyBanner(banner openrtb2.Banner) (*openrtb2.Banner, error) {
+	format := banner.Format
+
+	if banner.W == nil || banner.H == nil || *banner.W == 0 || *banner.H == 0 {
+		if len(format) == 0 {
+			return nil, &errortypes.BadInput{
+				Message: "Invalid size provided for Banner",
+			}
+		}
+
+		firstFormat := format[0]
+		banner.H = &firstFormat.H
```

Comment:
The imp is a copy, but the Banner field is a pointer to shared memory. You need to first make a shallow copy.